### PR TITLE
use charset consistently

### PR DESCRIPTION
**Pull request by yschimke**
I don't like this PR, so using it for discussion.

The spec says

> Encoding MIME Type: MIME Type for encoding of Data and Metadata. This MAY be a US-ASCII string that includes the Internet media type specified in RFC 2045. Many are registered with IANA such as CBOR. Suffix rules MAY be used for handling layout. For example, application/x.netflix+cbor or application/x.reactivesocket+cbor or application/x.netflix+json. The string may or may not be null terminated.

Does this actually mean:  "This MAY be a Internet media type specified in RFC 2045.  If present, the value MUST be US-ASCII.".  If it can be UTF-8 then string.length() may not be the byte length.

Also can we change the spec to say that the string MUST NOT be null terminated.  we need to string before creating strings.  I assume this is useful if we want to have 0-copy strings in C++.  But we create these garbage strings in Java anyway...

I'll add tests once this is agreed.

### CODE REVIEW

File: reactivesocket-core/src/main/java/io/reactivesocket/exceptions/Exceptions.java [hunk 1]
```diff
@@ -30,9 +30,15 @@ public static Throwable from(Frame frame) {
         final int errorCode = Frame.Error.errorCode(frame);
         String message = "<empty message>";
         final ByteBuffer byteBuffer = frame.getMetadata();
-        if (byteBuffer.hasArray()) {
```
Comment (yschimke) [inline reactivesocket-core/src/main/java/io/reactivesocket/exceptions/Exceptions.java]: hasArray here seems problematic in two ways.

1) assuming array backed
2) assuming that the byte buffer is not a sliced with a shared larger backing array

package io.reactivesocket.exceptions;

```
import java.nio.ByteBuffer;

public class AAAA {
    public static void main(String[] args) {
        ByteBuffer b = ByteBuffer.allocate(100);
        b.position(20).limit(50);
        ByteBuffer b2 = b.slice();

        System.out.println(b.capacity()); // 100
        System.out.println(b.remaining()); // 30
        System.out.println(b.array()); // [B@d716361
        System.out.println(b.array().length); // 100

        System.out.println();

        System.out.println(b2.capacity()); // 30
        System.out.println(b2.remaining()); // 30
        System.out.println(b2.array()); // [B@d716361
        System.out.println(b2.array().length); // 100
    }
}
```

**Comment by NiteshKant** [comment]
@yschimke I am not quiet sure if we want to use US-ASCII over UTF-8, do you want to open an issue for the protocol?

Since, mime-length is part of the frame, I am sure we don't intend it to be null terminated. It would help to make it clear in the spec.

File: reactivesocket-core/src/main/java/io/reactivesocket/exceptions/Exceptions.java (hunk 1 above)
Comment (NiteshKant) [inline reactivesocket-core/src/main/java/io/reactivesocket/exceptions/Exceptions.java]: Agreed, it is wrong to assume the buffer is array backed.

Comment (NiteshKant) [inline reactivesocket-core/src/main/java/io/reactivesocket/exceptions/Exceptions.java]: This code is defaulting to `<empty message>` if the buffer is not array backed, which isn't correct but is safe, rite?

Moreover, it is assuming `UTF-8` which isn't correct at all and it is converting entire metadata frame into the message. So, I guess ignoring error message if the buffer isn't array backed, isn't the bigger problem here 😄

**Comment by yschimke** [comment]
@NiteshKant I agree about using UTF-8, I'll fix the bug with string.length and remove the null termination handling.

**Comment by yschimke** [comment]
@NiteshKant anything blocking this merging? AFAIK it only fixes bug, no breaking behaviour.

**Comment by NiteshKant** [comment]
@yschimke just time, been busy today. I should be able to get to this today.

File: reactivesocket-core/src/main/java/io/reactivesocket/internal/frame/ByteBufferUtil.java [hunk 2]
```diff
@@ -44,4 +45,10 @@ public static ByteBuffer preservingSlice(final ByteBuffer byteBuffer, final int
         byteBuffer.limit(savedLimit).position(savedPosition);
         return result;
     }
+
+    public static String toUtfString(ByteBuffer byteBuffer) {
```
Comment (NiteshKant) [inline reactivesocket-core/src/main/java/io/reactivesocket/internal/frame/ByteBufferUtil.java]: nit: `toUtf8String`?

Comment (yschimke) [inline reactivesocket-core/src/main/java/io/reactivesocket/internal/frame/ByteBufferUtil.java]: fixed.  thanks

**Comment by stevegury** [comment]
👍 LGTM

**Pull request merged**